### DISCUSSION TITLE
view Apple-recognized objects?

### DISCUSSION THREAD

**Discussion by dmd** [RhetTbull/osxphotos#814] [Q&A]
Is there a way to see what objects Apple recognizes in a photo? For example, if I take a photo of a frog, I can search for 'frog' in Photos and it'll find that photo. I want to see what Photos is finding. (Bounding boxes for objects would be nice but not necessary).

**Comment by RhetTbull** [comment]
Yes. Osxphotos calls these "labels" and they're accessible through the `{label}` template, the `PhotoInfo.labels` property. You can also see them using the `osxphotos inspect` command which shows metadata for the photo currently selected in Photos. 

Bounding boxes are not supported and I'm not sure Photos stores these but it's been a while since I looked. The labels also include things which cannot be bounded such as location ("beach"), venue ("restaurant"), or holiday ("Christmas"). All this is available programmatically through the [search_info](https://github.com/RhetTbull/osxphotos/blob/master/API_README.md#searchinfo) property.

**Comment by dmd** [comment]
`labels` does not appear to be populated for me:

```
In [25]: len(photos)
Out[25]: 16036

In [26]: [p.search_info.labels for p in photos if p.search_info.labels]
Out[26]: []
```

And `Categories/Labels` shows up as `-` in inspect.

(But searching works just fine in Photos, e.g., if I search for 'beach' it shows me beach photos, even ones I have not manually tagged as such.)

**Reply by RhetTbull** [reply]
OS version?

**Reply by RhetTbull** [reply]
Also try running `osxphotos repl` and see if there's any warnings printed out regarding labels or search_info.

**Comment by dmd** [comment]
```
$ sw_vers
ProductName:		macOS
ProductVersion:		13.0
BuildVersion:		22A380

$ osxphotos repl
python version: 3.9.15 (main, Oct 11 2022, 21:39:54)
[Clang 14.0.0 (clang-1400.0.29.102)]
osxphotos version: 0.51.8
Using last opened Photos library: /Users/dmd/Pictures/Photos Library.photoslibrary
Loading database
Processing database /Users/dmd/Pictures/Photos Library.photoslibrary/database/photos.db
Processing database /Users/dmd/Pictures/Photos Library.photoslibrary/database/Photos.sqlite
Processing database.
Database version: 6000, 8.
Processing persons in photos.
Processing detected faces in photos.
Processing albums.
Processing keywords.
Processing photo details.
Processing import sessions.
Processing additional photo details.
Processing face details.
Processing photo labels.
Processing EXIF details.
Processing computed aesthetic scores.
Processing comments and likes for shared photos.
Processing moments.
Done processing details from Photos library.
Done: took 4.40 seconds
Getting photos
Found 16037 photos in 0.16 seconds

The following classes have been imported from osxphotos:
- AlbumInfo, ExifTool, PhotoInfo, PhotoExporter, ExportOptions, ExportResults, PhotosDB, PlaceInfo, QueryOptions, MomentInfo, ScoreInfo,
SearchInfo

The following variables are defined:
- photosdb: PhotosDB() instance for '/Users/dmd/Pictures/Photos Library.photoslibrary'
- photos: list of PhotoInfo objects for all photos filtered with any query options passed on command line (len=16037)
- all_photos: list of PhotoInfo objects for all photos in photosdb, including those in the trash (len=16094)
- selected: list of PhotoInfo objects for any photos selected in Photos (len=1)

The following functions may be helpful:
- get_photo(uuid): return a PhotoInfo object for photo with uuid; e.g. get_photo('B13F4485-94E0-41CD-AF71-913095D62E31')
- get_selected(); return list of PhotoInfo objects for photos selected in Photos
- show(photo): open a photo object in the default viewer; e.g. show(selected[0])
- show(path): open a file at path in the default viewer; e.g. show('/path/to/photo.jpg')
- spotlight(photo): open a photo and spotlight it in Photos
- inspect(object): print information about an object; e.g. inspect(PhotoInfo)
- explore(object): interactively explore an object with objexplore; e.g. explore(PhotoInfo)
- q, quit, quit(), exit, exit(): exit this interactive shell

>>> [p.search_info.labels for p in photos if p.search_info.labels]
[]

>>>
```

**Reply by RhetTbull** [reply]
Interesting, osxphotos didn't even attempt to load the search data (which includes labels).  Right click on your Photos library in Finder and select "Show package contents". Then look in the `database` folder. Do you have a a `search` sub folder? And if so, what are the contents?  osxphotos is expecting to find a `psi.sqlite` in `database/search` which contains the relevant details.

**Reply by dmd** [reply]
<img width="800" alt="image" src="https://user-images.githubusercontent.com/41439/200204735-746bb670-47e2-4a22-8faa-d6387412fcab.png">

**Reply by RhetTbull** [reply]
Thanks -- weird.  It appears the psi.sqlite *is* being loaded (it happens right after this `Processing face details.`) but that the data is not being associated correctly with photos.  If there was an error parsing the data then osxphotos would log a warning which doesn't appear.  

My guess is the mapping of UUID in the psi.sqlite to UUID in the Photos.sqlite isn't working correctly.  In earlier versions of macOS, the UUID for the associated photos is stored as two INTs in the psi.sqlite (I have no idea why as it seems it would be more efficient to have a single primary key (UUID) than two).  osxphotos combines these two ints to produce the UUID for the photo which maps to `Photos.sqlite`.  The code that does the conversion is here:

https://github.com/RhetTbull/osxphotos/blob/2f9b0b8fa18995684a73cfa010989253dc79be16/osxphotos/photosdb/_photosdb_process_searchinfo.py#L196-L206

The ints are stored as little endian signed ints of lenght 8 bytes.  According to [this](https://developer.apple.com/documentation/apple-silicon/porting-your-macos-apps-to-apple-silicon) both Apple Silicon and x86 use little endian format so that shouldn't be an issue. 

I did a [diff of psi.sqlite](https://github.com/RhetTbull/osxphotos/wiki/Photos-search-data-(psi.sqlite)-diff-Monterey-to-Ventura) from 12.0 to 13.0 and there's a couple of differences in schema but nothing in the columns used by osxphotos.

I don't have access macOS 13 machine so I can't test this directly.  I'll try to think about adding some debugging in to help figure this out.  In the mean time, if you don't mind creating a sample library that you could share, that would be helpful.  If you hold down `Option` while starting Photos you can create a new empty photos library. Then copy a few photos to the new library that should have easily identifiable labels (e.g. dog, cat, beach, etc). Allow the computer to sit overnight with Photos in the background then verify you can search for the appropriate terms in Photos. Then send me the zipped photos library. (rturnbull+git@gmail.com).  The AI/ML photo analysis only happens when Photos isn't being used, hence the need to wait a few hours for `photoanalysisd` to do its thing.

**Reply by dmd** [reply]
Sent.

**Reply by RhetTbull** [reply]
Got the test library -- thanks! I figured out the problem.  The `psi.sqlite` file is where Photos stores all the ML analysis of each photo (including labels/categories) and the `groups.category` column indicates what the data is.  For earlier versions of Photos, the labels have `category = 2024` (determined by reverse engineering) but in your macOS 13.0 database, the labels have `category = 1500` thus osxphotos didn't detect this data as labels.  This wasn't caught in tests because the test library I have for 13.0 was one someone in the r/macos subreddit provided for testing and it didn't have any labels--this is expected when the library hasn't been in existence long enough for `photoanalysisd` to do its thing. 

I'll need to do some re-engineering to allow osxphotos to lookup the correct categories based on macOS version.  In the meantime, the following code will allow you to inspect the labels.

```python
import typing as t

import osxphotos

SEARCH_CATEGORY_LABEL = 1500


def get_photo_labels(photo: osxphotos.PhotoInfo) -> t.List[str]:
    """Return list of normalized search info labels for photo"""
    try:
        search_record = photo._db._db_searchinfo_uuid[photo.uuid]
    except KeyError:
        return []
    return [
        rec["normalized_string"]
        for rec in search_record
        if rec["category"] == SEARCH_CATEGORY_LABEL
    ]
```

```pycon
osxphotos repl --db testlibrary_13_0.photoslibrary
python version: 3.9.13 (main, Nov  6 2022, 18:33:56)
[Clang 12.0.0 (clang-1200.0.32.29)]
osxphotos version: 0.52.0
Loading database
Processing database /Users/rhet/Dropbox/Code/osxphotos/testlibrary_13_0.photoslibrary/database/photos.db
Processing database /Users/rhet/Dropbox/Code/osxphotos/testlibrary_13_0.photoslibrary/database/Photos.sqlite
Processing database.
Database version: 6000, 8.
Processing persons in photos.
Processing detected faces in photos.
Processing albums.
Processing keywords.
Processing photo details.
Processing import sessions.
Processing additional photo details.
Processing face details.
Processing photo labels.
Processing EXIF details.
Processing computed aesthetic scores.
Processing comments and likes for shared photos.
Processing moments.
Done processing details from Photos library.
Done: took 0.15 seconds
Getting photos
Found 4 photos in 0.00 seconds

>>> from labels import get_photo_labels
>>> for p in photos:
...     print(p.uuid, get_photo_labels(p))
34F18324-B20D-47C1-9858-136C7A6F585F
['fence', 'people', 'palm tree', 'tree', 'foliage', 'plant']
2CE9BAC2-7E61-4747-9C20-43337AA48B90
['rocks', 'people', 'fence']
B150DE83-3EDF-496A-A028-0504800A12F1
['hoodie', 'clothing', 'outdoor', 'land', 'beach', 'water', 'water body', 'ocean', 'people', 'child']
F6221127-3F65-40E5-9038-DBBD8CDA1D6D
['people', 'child', 'rocks']
>>>
```

**Reply by dmd** [reply]
Can you give me a rough idea of when you plan to do this - like is this something you'll do in the next few days, weeks, months? Only asking to know if it's worth my writing a workaround to 'inspect' right now, or if it's not worth the bother :)

**Reply by RhetTbull** [reply]
Hopefully within a week or so. This is a part time thing for me that I do mostly on weekends. Fixing the labels is relatively easy and can definitely get that done in a week or two max. All the other search categories in `SearchInfo` will take more work as I've  got check each of them to verify they didn't change.

**Comment by dmd** [comment]
Awesome. BTW if you enable the 'sponsor' button on this I'd happily do it,
osxphotos is massively useful to my workflow!

Daniel
